src/core/Song.cpp

```diff
@@ -1275,6 +1276,159 @@ void Song::exportProject(bool multiExport)
 }
 
 
+#if 0
+static void traverseDomElement(const QDomElement &e, int level=0) {
+	
+}
+#endif
+
+void Song::exportProjectMidi()
+{
+	if( isEmpty() )
+	{
+		QMessageBox::information( gui->mainWindow(),
+				tr( "Empty project" ),
+				tr( "This project is empty so exporting makes "
+					"no sense. Please put some items into "
+					"Song Editor first!" ) );
+		return;
+	}
+
+	FileDialog efd( gui->mainWindow() );
+	
+	efd.setFileMode( FileDialog::AnyFile );
+	
+	QStringList types;
+	types << tr("MIDI File (*.mid)");
+	efd.setNameFilters( types );
+	QString base_filename;
+	if( !m_fileName.isEmpty() )
+	{
+		efd.setDirectory( QFileInfo( m_fileName ).absolutePath() );
+		base_filename = QFileInfo( m_fileName ).completeBaseName();
+	}
+	else
+	{
+		efd.setDirectory( ConfigManager::inst()->userProjectsDir() );
+		base_filename = tr( "untitled" );
+	}
+	efd.selectFile( base_filename + ".mid" );
+	efd.setWindowTitle( tr( "Select file for project-export..." ) );
+
+	efd.setAcceptMode( FileDialog::AcceptSave );
+
+
+	if( efd.exec() == QDialog::Accepted && !efd.selectedFiles().isEmpty() && !efd.selectedFiles()[0].isEmpty() )
+	{
+		QString suffix = "";
+		int stx = efd.selectedNameFilter().indexOf( "(*." );
+		int etx = efd.selectedNameFilter().indexOf( ")" );
+
+		if ( stx > 0 && etx > stx ) 
+		{
+			// Get first extension from selected dropdown.
+			// i.e. ".wav" from "WAV-File (*.wav), Dummy-File (*.dum)"
+			suffix = efd.selectedNameFilter().mid( stx + 2, etx - stx - 2 ).split( " " )[0].trimmed();
+			if ( efd.selectedFiles()[0].endsWith( suffix ) )
+			{
+				suffix = "";
+			}
+		}
+
+		const QString export_file_name = efd.selectedFiles()[0] + suffix;
+		
+		// NOTE start midi export
+		QFile f(export_file_name);
+		f.open(QIODevice::WriteOnly);
+		QDataStream midiout(&f);
+		
+		TrackContainer::TrackList tracks;
+		InstrumentTrack* instTrack;
+		QDomElement element;
+	
+
+		tracks += Engine::getSong()->tracks();
+		tracks += Engine::getBBTrackContainer()->tracks();
+ 		int nTracks = 0;
+		const int BUFFER_SIZE = 50*1024;
+		uint8_t buffer[BUFFER_SIZE];
+		uint32_t size;
+		
+		foreach( Track* track, tracks ) if( track->type() == Track::InstrumentTrack ) nTracks++;
+		
+		// midi header
+		MidiFile::MIDIHeader header(nTracks);
+		size = header.writeToBuffer(buffer);
+		midiout.writeRawData((char *)buffer, size);
+		
+		// midi tracks
+		foreach( Track* track, tracks )
```

Comment:
Do you think we can modularize this so that we are able to export individual tracks in the future?